src/vs/base/node/powershell.ts

```diff
@@ -0,0 +1,331 @@
+
+// The platform details descriptor for the platform we're on
+const isProcess64Bit: boolean = process.arch === 'x64';
+const isOS64Bit: boolean = isProcess64Bit || env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+
+interface IPowerShellExeDetails {
+	readonly displayName: string;
+	readonly exePath: string;
+}
+
+interface IPossiblePowerShellExe extends IPowerShellExeDetails {
+	exists(): boolean;
+}
+
+class PossiblePowerShellExe implements IPossiblePowerShellExe {
+	public readonly exePath: string;
+	public readonly displayName: string;
+
+	private knownToExist: boolean | undefined;
+
+	constructor(
+		pathToExe: string,
+		installationName: string,
+		{ knownToExist = false }: { knownToExist?: boolean } = {}) {
+
+		this.exePath = pathToExe;
+		this.displayName = installationName;
+		this.knownToExist = knownToExist || undefined;
+	}
+
+	public exists(): boolean {
+		if (this.knownToExist === undefined) {
+			this.knownToExist = fs.existsSync(this.exePath);
+		}
+		return this.knownToExist;
+	}
+}
+
+function getProgramFilesPath(
+	{ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string | null {
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return env.ProgramFiles!;
+	}
+
+	// We might be a 64-bit process looking for 32-bit program files
+	if (isProcess64Bit) {
+		return env['ProgramFiles(x86)']!;
+	}
+
+	// We might be a 32-bit process looking for 64-bit program files
+	if (isOS64Bit) {
+		return env.ProgramW6432!;
+	}
+
+	// We're a 32-bit process on 32-bit Windows, there is no other Program Files dir
+	return null;
+}
+
+function getSystem32Path({ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string {
+	const windir: string = env.windir!;
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return path.join(windir, 'System32');
+	}
+
+	// We might be a 64-bit process looking for 32-bit system32
+	if (isProcess64Bit) {
+		return path.join(windir, 'SysWOW64');
+	}
+
+	// We might be a 32-bit process looking for 64-bit system32
+	if (isOS64Bit) {
+		return path.join(windir, 'Sysnative');
+	}
+
+	// We're on a 32-bit Windows, so no alternate bitness
+	return path.join(windir, 'System32');
+}
+
+function findPSCoreWindowsInstallation(
+	{ useAlternateBitness = false, findPreview = false }:
+		{ useAlternateBitness?: boolean; findPreview?: boolean } = {}): IPossiblePowerShellExe | null {
+
+	const programFilesPath = getProgramFilesPath({ useAlternateBitness });
+
+	if (!programFilesPath) {
+		return null;
+	}
+
+	const powerShellInstallBaseDir = path.join(programFilesPath, 'PowerShell');
+
+	// Ensure the base directory exists
+	if (!(fs.existsSync(powerShellInstallBaseDir) && fs.lstatSync(powerShellInstallBaseDir).isDirectory())) {
+		return null;
+	}
+
+	let highestSeenVersion: number = -1;
+	let pwshExePath: string | null = null;
+	for (const item of fs.readdirSync(powerShellInstallBaseDir)) {
```

Comment:
Suggestion: This loop is a candidate to be moved into a small function whose purpose is to grab and return the pwshExePath variable?

Reply:
I could... but that's most of this function. I'm not sure that buys a whole lot of readability here IMO. If more folks think I should do this, I'm happy to be out voted.